File: gwen/lexer.py
"""Gwen lexer - tokenizes .gw source files."""

from enum import Enum, auto
from dataclasses import dataclass
from typing import List


class TokenType(Enum):
    # Literals
    INTEGER = auto()
    FLOAT = auto()
    STRING = auto()
    IDENTIFIER = auto()
    TAG = auto()           # @name
    AS = auto()            # as

    # Keywords
    FUNC = auto()
    ENDFUNC = auto()
    IF = auto()
    THEN = auto()
    ELIF = auto()
    ELSE = auto()
    ENDIF = auto()
    WHILE = auto()
    DO = auto()
    ENDWHILE = auto()
    FOR = auto()
    IN = auto()
    TO = auto()
    STEP = auto()
    ORDER = auto()       # order (强制正序)
    REVERSE = auto()     # reverse (强制倒序)
    WITH = auto()
    INDEX = auto()
    ENDFOR = auto()
    MATCH = auto()
    WHEN = auto()
    ENDMATCH = auto()
    MODULE = auto()
    ENDMODULE = auto()
    EXPORT = auto()
    USE = auto()
    FROM = auto()
    RETURN = auto()
    PARALLEL = auto()
    ENDPARALLEL = auto()
    ALLOWFAIL = auto()
    OK = auto()
    ERR = auto()
    AND = auto()
    OR = auto()
    NOT = auto()
    TRUE = auto()
    FALSE = auto()
    MOD = auto()
    GLOBAL = auto()        # global
    CONST = auto()         # const
    ARENA = auto()         # arena
    ENDARENA = auto()      # endarena
    VAR = auto()           # var
    ENDVAR = auto()        # endvar
    DEFAULT = auto()       # default

    # Operators
    ASSIGN = auto()        # :=
    ARROW = auto()         # ->
    FAT_ARROW = auto()     # =>
    EQ = auto()            # =
    NEQ = auto()           # !=
    LT = auto()            # <
    GT = auto()            # >
    LTE = auto()           # <=
    GTE = auto()           # >=
    PLUS = auto()          # +
    MINUS = auto()         # -
    STAR = auto()          # *
    SLASH = auto()         # /
    CARET = auto()          # ^

    # Delimiters
    LPAREN = auto()        # (
    RPAREN = auto()        # )
    LBRACKET = auto()      # [
    RBRACKET = auto()      # ]
    COMMA = auto()         # ,
    COLON = auto()         # :
    DOT = auto()           # .
    NEWLINE = auto()

    # Special
    EOF = auto()


KEYWORDS = {
    "func": TokenType.FUNC,
    "endfunc": TokenType.ENDFUNC,
    "if": TokenType.IF,
    "then": TokenType.THEN,
    "elif": TokenType.ELIF,
    "else": TokenType.ELSE,
    "endif": TokenType.ENDIF,
    "while": TokenType.WHILE,
    "do": TokenType.DO,
    "endwhile": TokenType.ENDWHILE,
    "for": TokenType.FOR,
    "in": TokenType.IN,
    "to": TokenType.TO,
    "step": TokenType.STEP,
    "with": TokenType.WITH,
    "index": TokenType.INDEX,
    "endfor": TokenType.ENDFOR,
    "match": TokenType.MATCH,
    "when": TokenType.WHEN,
    "endmatch": TokenType.ENDMATCH,
    "module": TokenType.MODULE,
    "endmodule": TokenType.ENDMODULE,
    "export": TokenType.EXPORT,
    "use": TokenType.USE,
    "from": TokenType.FROM,
    "return": TokenType.RETURN,
    "parallel": TokenType.PARALLEL,
    "endparallel": TokenType.ENDPARALLEL,
    "allowfail": TokenType.ALLOWFAIL,
    "ok": TokenType.OK,
    "err": TokenType.ERR,
    "as": TokenType.AS,
    "and": TokenType.AND,
    "or": TokenType.OR,
    "not": TokenType.NOT,
    "true": TokenType.TRUE,
    "false": TokenType.FALSE,
    "mod": TokenType.MOD,
    "global": TokenType.GLOBAL,
    "const": TokenType.CONST,
    "arena": TokenType.ARENA,
    "endarena": TokenType.ENDARENA,
    "var": TokenType.VAR,
    "endvar": TokenType.ENDVAR,
    "default": TokenType.DEFAULT,
    "order": TokenType.ORDER,
    "reverse": TokenType.REVERSE,
}


@dataclass
class Token:
    type: TokenType
    value: str
    line: int
    column: int

    def __repr__(self):
        return f"Token({self.type.name}, {self.value!r}, L{self.line}:{self.column})"


class LexerError(Exception):
    def __init__(self, message: str, line: int, column: int):
        super().__init__(f"Lexer error at L{line}:{column}: {message}")
        self.line = line
        self.column = column


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []

    def peek(self) -> str:
        if self.pos >= len(self.source):
            return "\0"
        return self.source[self.pos]

    def peek_next(self) -> str:
        if self.pos + 1 >= len(self.source):
            return "\0"
        return self.source[self.pos + 1]

    def advance(self) -> str:
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.column = 1
        else:
            self.column += 1
        return ch

    def add_token(self, token_type: TokenType, value: str, line: int, column: int):
        self.tokens.append(Token(token_type, value, line, column))

    def skip_whitespace(self):
        while self.pos < len(self.source) and self.peek() in (" ", "\t", "\r"):
            self.advance()

    def skip_line_comment(self):
        # // single line comment - consume until newline
        while self.pos < len(self.source) and self.peek() != "\n":
            self.advance()

    def skip_block_comment(self):
        # /* ... */ block comment (not nested)
        start_line, start_col = self.line, self.column
        self.advance()  # consume /
        self.advance()  # consume *
        while self.pos < len(self.source):
            if self.peek() == "*" and self.peek_next() == "/":
                self.advance()  # consume *
                self.advance()  # consume /
                return
            self.advance()
        raise LexerError("Unterminated block comment", start_line, start_col)

    def read_string(self):
        line, col = self.line, self.column
        quote = self.advance()  # consume opening quote
        result = ""
        while self.pos < len(self.source) and self.peek() != quote:
            if self.peek() == "\\":
                self.advance()
                ch = self.advance()
                if ch == "n":
                    result += "\n"
                elif ch == "t":
                    result += "\t"
                elif ch == "\\":
                    result += "\\"
                elif ch == quote:
                    result += quote
                else:
                    result += "\\" + ch
            else:
                result += self.advance()
        if self.pos >= len(self.source):
            raise LexerError("Unterminated string", line, col)
        self.advance()  # consume closing quote
        self.add_token(TokenType.STRING, result, line, col)

    def read_number(self):
        line, col = self.line, self.column
        result = ""
        is_float = False
        while self.pos < len(self.source) and (self.peek().isdigit() or self.peek() == "."):
            if self.peek() == ".":
                if is_float:
                    break
                is_float = True
            result += self.advance()
        token_type = TokenType.FLOAT if is_float else TokenType.INTEGER
        self.add_token(token_type, result, line, col)

    def read_identifier_or_keyword(self):
        line, col = self.line, self.column
        result = ""
        while self.pos < len(self.source) and (self.peek().isalnum() or self.peek() == "_"):
            result += self.advance()
        token_type = KEYWORDS.get(result, TokenType.IDENTIFIER)
        self.add_token(token_type, result, line, col)

    def read_tag(self):
        line, col = self.line, self.column
        self.advance()  # consume @
        name = ""
        while self.pos < len(self.source) and (self.peek().isalnum() or self.peek() == "_"):
            name += self.advance()
        if not name:
            raise LexerError("Expected tag name after @", line, col)
        self.add_token(TokenType.TAG, name, line, col)

    def tokenize(self) -> List[Token]:
        while self.pos < len(self.source):
            ch = self.peek()

            # Whitespace (not newline)
            if ch in (" ", "\t", "\r"):
                self.skip_whitespace()
                continue

            # Newline
            if ch == "\n":
                line, col = self.line, self.column
                self.advance()
                # Collapse multiple newlines
                if not self.tokens or self.tokens[-1].type != TokenType.NEWLINE:
                    self.add_token(TokenType.NEWLINE, "\\n", line, col)
                continue

            # Comment
            # Line comment
            if ch == "/" and self.peek_next() == "/":
                self.skip_line_comment()
                continue

            # Block comment
            if ch == "/" and self.peek_next() == "*":
                self.skip_block_comment()
                continue

            # String
            if ch in ('"', "'"):
                self.read_string()
                continue

            # Number
            if ch.isdigit():
                self.read_number()
                continue

            # Identifier or keyword
            if ch.isalpha() or ch == "_":
                self.read_identifier_or_keyword()
                continue

            # Tag
            if ch == "@":
                self.read_tag()
                continue

            # Two-character operators
            line, col = self.line, self.column
            if ch == ":" and self.peek_next() == "=":
                self.advance()
                self.advance()
                self.add_token(TokenType.ASSIGN, ":=", line, col)
                continue
            if ch == "-" and self.peek_next() == ">":
                self.advance()
                self.advance()
                self.add_token(TokenType.ARROW, "->", line, col)
                continue
            if ch == "=" and self.peek_next() == ">":
                self.advance()
                self.advance()
                self.add_token(TokenType.FAT_ARROW, "=>", line, col)
                continue
            if ch == "!" and self.peek_next() == "=":
                self.advance()
                self.advance()
                self.add_token(TokenType.NEQ, "!=", line, col)
                continue
            if ch == "<" and self.peek_next() == "=":
                self.advance()
                self.advance()
                self.add_token(TokenType.LTE, "<=", line, col)
                continue
            if ch == ">" and self.peek_next() == "=":
                self.advance()
                self.advance()
                self.add_token(TokenType.GTE, ">=", line, col)
                continue


            # Single-character operators/delimiters
            single_chars = {
                "=": TokenType.EQ,
                "<": TokenType.LT,
                ">": TokenType.GT,
                "+": TokenType.PLUS,
                "-": TokenType.MINUS,
                "*": TokenType.STAR,
                "/": TokenType.SLASH,
                "^": TokenType.CARET,
                "(": TokenType.LPAREN,
                ")": TokenType.RPAREN,
                "[": TokenType.LBRACKET,
                "]": TokenType.RBRACKET,
                ",": TokenType.COMMA,
                ":": TokenType.COLON,
                ".": TokenType.DOT,
            }

            if ch in single_chars:
                self.advance()
                self.add_token(single_chars[ch], ch, line, col)
                continue

            raise LexerError(f"Unexpected character: {ch!r}", self.line, self.column)

        self.add_token(TokenType.EOF, "", self.line, self.column)
        return self.tokens


def tokenize(source: str) -> List[Token]:
    """Convenience function to tokenize source code."""
    return Lexer(source).tokenize()

File: internal/lexer/lexer.go
package lexer

import (
	"fmt"
	"strings"
	"unicode"

	"github.com/Cass-ette/gwen-lang/internal/token"
)

const eofRune = rune(0)

type Error struct {
	Message string
	Line    int
	Column  int
}

func (e *Error) Error() string {
	return fmt.Sprintf("lexer error at L%d:%d: %s", e.Line, e.Column, e.Message)
}

type Lexer struct {
	source []rune
	pos    int
	line   int
	column int
	tokens []token.Token
}

func New(source string) *Lexer {
	return &Lexer{
		source: []rune(source),
		line:   1,
		column: 1,
	}
}

func Tokenize(source string) ([]token.Token, error) {
	return New(source).Tokenize()
}

func (l *Lexer) Tokenize() ([]token.Token, error) {
	for l.pos < len(l.source) {
		ch := l.peek()

		if ch == ' ' || ch == '\t' || ch == '\r' {
			l.skipWhitespace()
			continue
		}

		if ch == '\n' {
			line, column := l.line, l.column
			l.advance()
			if len(l.tokens) == 0 || l.tokens[len(l.tokens)-1].Type != token.Newline {
				l.addToken(token.Newline, "\n", line, column)
			}
			continue
		}

		if ch == '/' && l.peekNext() == '/' {
			l.skipLineComment()
			continue
		}

		if ch == '/' && l.peekNext() == '*' {
			if err := l.skipBlockComment(); err != nil {
				return nil, err
			}
			continue
		}

		if ch == '"' || ch == '\'' {
			if err := l.readString(); err != nil {
				return nil, err
			}
			continue
		}

		if unicode.IsDigit(ch) {
			l.readNumber()
			continue
		}

		if unicode.IsLetter(ch) || ch == '_' {
			l.readIdentifierOrKeyword()
			continue
		}

		if ch == '@' {
			if err := l.readTag(); err != nil {
				return nil, err
			}
			continue
		}

		line, column := l.line, l.column

		switch {
		case ch == ':' && l.peekNext() == '=':
			l.advance()
			l.advance()
			l.addToken(token.Assign, ":=", line, column)
			continue
		case ch == '-' && l.peekNext() == '>':
			l.advance()
			l.advance()
			l.addToken(token.Arrow, "->", line, column)
			continue
		case ch == '=' && l.peekNext() == '>':
			l.advance()
			l.advance()
			l.addToken(token.FatArrow, "=>", line, column)
			continue
		case ch == '!' && l.peekNext() == '=':
			l.advance()
			l.advance()
			l.addToken(token.Neq, "!=", line, column)
			continue
		case ch == '<' && l.peekNext() == '=':
			l.advance()
			l.advance()
			l.addToken(token.Lte, "<=", line, column)
			continue
		case ch == '>' && l.peekNext() == '=':
			l.advance()
			l.advance()
			l.addToken(token.Gte, ">=", line, column)
			continue
		}

		if tokenType, ok := singleCharTokens[ch]; ok {
			l.advance()
			l.addToken(tokenType, string(ch), line, column)
			continue
		}

		return nil, &Error{
			Message: fmt.Sprintf("unexpected character: %q", ch),
			Line:    l.line,
			Column:  l.column,
		}
	}

	l.addToken(token.EOF, "", l.line, l.column)
	return l.tokens, nil
}

var singleCharTokens = map[rune]token.Type{
	'=': token.Eq,
	'<': token.Lt,
	'>': token.Gt,
	'+': token.Plus,
	'-': token.Minus,
	'*': token.Star,
	'/': token.Slash,
	'^': token.Caret,
	'(': token.LParen,
	')': token.RParen,
	'[': token.LBracket,
	']': token.RBracket,
	'{': token.LBrace,
	'}': token.RBrace,
	',': token.Comma,
	':': token.Colon,
	'.': token.Dot,
}

func (l *Lexer) peek() rune {
	if l.pos >= len(l.source) {
		return eofRune
	}
	return l.source[l.pos]
}

func (l *Lexer) peekNext() rune {
	if l.pos+1 >= len(l.source) {
		return eofRune
	}
	return l.source[l.pos+1]
}

func (l *Lexer) advance() rune {
	ch := l.source[l.pos]
	l.pos++
	if ch == '\n' {
		l.line++
		l.column = 1
	} else {
		l.column++
	}
	return ch
}

func (l *Lexer) addToken(tokenType token.Type, value string, line, column int) {
	l.tokens = append(l.tokens, token.Token{
		Type:   tokenType,
		Value:  value,
		Line:   line,
		Column: column,
	})
}

func (l *Lexer) skipWhitespace() {
	for l.pos < len(l.source) {
		ch := l.peek()
		if ch != ' ' && ch != '\t' && ch != '\r' {
			return
		}
		l.advance()
	}
}

func (l *Lexer) skipLineComment() {
	for l.pos < len(l.source) && l.peek() != '\n' {
		l.advance()
	}
}

func (l *Lexer) skipBlockComment() error {
	startLine, startColumn := l.line, l.column
	l.advance()
	l.advance()

	for l.pos < len(l.source) {
		if l.peek() == '*' && l.peekNext() == '/' {
			l.advance()
			l.advance()
			return nil
		}
		l.advance()
	}

	return &Error{
		Message: "unterminated block comment",
		Line:    startLine,
		Column:  startColumn,
	}
}

func (l *Lexer) readString() error {
	line, column := l.line, l.column
	quote := l.advance()
	var builder strings.Builder

	for l.pos < len(l.source) && l.peek() != quote {
		if l.peek() == '\\' {
			l.advance()
			if l.pos >= len(l.source) {
				return &Error{
					Message: "unterminated string",
					Line:    line,
					Column:  column,
				}
			}

			ch := l.advance()
			switch ch {
			case 'n':
				builder.WriteRune('\n')
			case 't':
				builder.WriteRune('\t')
			case '\\':
				builder.WriteRune('\\')
			default:
				if ch == quote {
					builder.WriteRune(quote)
				} else {
					builder.WriteRune('\\')
					builder.WriteRune(ch)
				}
			}
			continue
		}

		builder.WriteRune(l.advance())
	}

	if l.pos >= len(l.source) {
		return &Error{
			Message: "unterminated string",
			Line:    line,
			Column:  column,
		}
	}

	l.advance()
	l.addToken(token.String, builder.String(), line, column)
	return nil
}

func (l *Lexer) readNumber() {
	line, column := l.line, l.column
	var builder strings.Builder
	isFloat := false

	for l.pos < len(l.source) {
		ch := l.peek()
		if !unicode.IsDigit(ch) && ch != '.' {
			break
		}
		if ch == '.' {
			if isFloat {
				break
			}
			isFloat = true
		}
		builder.WriteRune(l.advance())
	}

	tokenType := token.Integer
	if isFloat {
		tokenType = token.Float
	}
	l.addToken(tokenType, builder.String(), line, column)
}

func (l *Lexer) readIdentifierOrKeyword() {
	line, column := l.line, l.column
	var builder strings.Builder

	for l.pos < len(l.source) {
		ch := l.peek()
		if !unicode.IsLetter(ch) && !unicode.IsDigit(ch) && ch != '_' {
			break
		}
		builder.WriteRune(l.advance())
	}

	value := builder.String()
	l.addToken(token.LookupIdentifier(value), value, line, column)
}

func (l *Lexer) readTag() error {
	line, column := l.line, l.column
	l.advance()

	var builder strings.Builder
	for l.pos < len(l.source) {
		ch := l.peek()
		if !unicode.IsLetter(ch) && !unicode.IsDigit(ch) && ch != '_' {
			break
		}
		builder.WriteRune(l.advance())
	}

	if builder.Len() == 0 {
		return &Error{
			Message: "expected tag name after @",
			Line:    line,
			Column:  column,
		}
	}

	l.addToken(token.Tag, builder.String(), line, column)
	return nil
}
